Number of REST fixes for Altair.

cheatfate opened the pull request.

cheatfate committed:
Fix getForkSchedule call.
Create cache of all configuration endpoints at node startup.
Add prepareJsonResponse() call to create cached responses.
Mark all procedures with `raises`.

cheatfate committed:
Add getForkSchedule to VC.
Fix getForkSchedule return type for API.
More `raises` annotations.
Fix VC fork_service.nim.

arnetheduck commented on beacon_chain/spec/eth2_apis/eth2_rest_serialization.nim:
```diff
@@ -58,115 +58,193 @@ type
+
+proc jsonResponseWRoot*(t: typedesc[RestApiResponse], data: auto,
+                        dependent_root: Eth2Digest): RestApiResponse {.
+     raises: [Defect].} =
```
why not a `push raises defect`?

cheatfate replied:
I'm just not sure that `{.async.}` procedures could be annotated with `Defect` only.

cheatfate replied:
Fixed.

arnetheduck commented on beacon_chain/validator_client/fork_service.nim:
```diff
@@ -1,23 +1,70 @@
+              err_name = exc.name, err_msg = exc.msg
+        return
+
+    let fork =
```
so.. if we get a fork here that hasn't happened yet, this code will keep polling?

cheatfate replied:
Yes it will keep polling it every epoch.

cheatfate committed:
Use `push raises` instead of inline `raises`.

cheatfate committed:
Improvements for REST API aggregated attestations and attestations processing.

cheatfate committed:
Rename eth2_network.sendXXX procedures to eth2_network.broadcastXXX.
Add broadcastBeaconBlock() and broadcastAggregateAndProof().
Fix links to specification in REST API declarations.
Add implementation for v2 getStateV2().
Add validator_duties.sendXXX procedures which not only broadcast data, but also validate it.
Fix JSON-RPC/REST to use new validator_duties.sendXXX procedures instead of own implementations.

arnetheduck commented on beacon_chain/rpc/rest_beacon_api.nim:
```diff
@@ -788,21 +765,71 @@ proc installBeaconApiHandlers*(router: var RestRouter, node: BeaconNode) =
+      block:
+        var res: seq[Future[void]]
+        for attestation in attestations:
+          res.add(processAttestation(attestation))
```
sendAttestation?

cheatfate committed:
Fix validator_client online nodes count incorrect value.
Fix aggregate and proof attestation could be sent too late.

cheatfate committed:
Adding timeout for block wait in attestations processing.
Fix compilation errors.

cheatfate committed:
Attempt to debug aggregate and proofs.

cheatfate committed:
Fix Beacon AIP to use `sendAttestation`.
Add link comment to produceBlockV2.

cheatfate committed:
Add debug logs before publish operation for blocks, attestations and aggregated attestations.
Fix attestations publishing issue.

arnetheduck committed:
logging fixes

`indexInCommnittee` already logged in attestation
